File: apps/desktop/src/stores/workflow.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

export interface Workflow {
  id: string
  name: string
  description?: string
  status: 'draft' | 'active' | 'paused' | 'archived'
  nodes: any[]
  edges: any[]
  createdAt: string
  updatedAt: string
  lastRunAt?: string
  runCount: number
}

export interface WorkflowRun {
  id: string
  workflowId: string
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'
  startedAt: string
  completedAt?: string
  inputs: Record<string, any>
  outputs: Record<string, any>
  error?: string
}

export const useWorkflowStore = defineStore('workflow', () => {
  const workflows = ref<Workflow[]>([])
  const currentWorkflow = ref<Workflow | null>(null)
  const currentRun = ref<WorkflowRun | null>(null)
  const isLoading = ref(false)
  const isRunning = ref(false)
  const error = ref<string | null>(null)

  const activeWorkflows = computed(() => 
    workflows.value.filter(w => w.status === 'active')
  )
  const draftWorkflows = computed(() => 
    workflows.value.filter(w => w.status === 'draft')
  )
  const totalWorkflows = computed(() => workflows.value.length)

  function getWorkflow(id: string): Workflow | undefined {
    return workflows.value.find(w => w.id === id)
  }

  function addWorkflow(workflow: Workflow): void {
    const existing = getWorkflow(workflow.id)
    if (existing) {
      Object.assign(existing, workflow)
    } else {
      workflows.value.push(workflow)
    }
  }

  function removeWorkflow(id: string): void {
    const index = workflows.value.findIndex(w => w.id === id)
    if (index !== -1) {
      workflows.value.splice(index, 1)
    }
    if (currentWorkflow.value?.id === id) {
      currentWorkflow.value = null
    }
  }

  function updateWorkflow(id: string, updates: Partial<Workflow>): void {
    const workflow = getWorkflow(id)
    if (workflow) {
      Object.assign(workflow, updates, { updatedAt: new Date().toISOString() })
    }
  }

  function selectWorkflow(workflow: Workflow | null): void {
    currentWorkflow.value = workflow
  }

  function createNewWorkflow(): Workflow {
    const workflow: Workflow = {
      id: `workflow-${Date.now()}`,
      name: 'New Workflow',
      status: 'draft',
      nodes: [],
      edges: [],
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
      runCount: 0
    }
    addWorkflow(workflow)
    return workflow
  }

  function duplicateWorkflow(id: string): Workflow | null {
    const original = getWorkflow(id)
    if (!original) return null

    const duplicate: Workflow = {
      ...original,
      id: `workflow-${Date.now()}`,
      name: `${original.name} (Copy)`,
      status: 'draft',
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
      runCount: 0
    }
    addWorkflow(duplicate)
    return duplicate
  }

  function setCurrentRun(run: WorkflowRun | null): void {
    currentRun.value = run
  }

  function clearWorkflows(): void {
    workflows.value = []
    currentWorkflow.value = null
    currentRun.value = null
  }

  return {
    workflows,
    currentWorkflow,
    currentRun,
    isLoading,
    isRunning,
    error,
    activeWorkflows,
    draftWorkflows,
    totalWorkflows,
    getWorkflow,
    addWorkflow,
    removeWorkflow,
    updateWorkflow,
    selectWorkflow,
    createNewWorkflow,
    duplicateWorkflow,
    setCurrentRun,
    clearWorkflows
  }
})

export default useWorkflowStore

File: apps/server/app/schemas/workflow.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from enum import Enum
from datetime import datetime


class WorkflowStatus(str, Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    PAUSED = "paused"
    ARCHIVED = "archived"


class WorkflowRunStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class WorkflowNodeType(str, Enum):
    START = "start"
    END = "end"
    CONDITION = "condition"
    PARALLEL = "parallel"
    MERGE = "merge"
    DELAY = "delay"
    PDF_MERGE = "pdf-merge"
    PDF_SPLIT = "pdf-split"
    PDF_CONVERT = "pdf-convert"
    PDF_COMPRESS = "pdf-compress"
    PDF_WATERMARK = "pdf-watermark"
    PDF_ENCRYPT = "pdf-encrypt"
    PDF_DECRYPT = "pdf-decrypt"
    OCR_PROCESS = "ocr-process"
    AI_SUMMARY = "ai-summary"
    AI_TRANSLATE = "ai-translate"
    AI_EXTRACT = "ai-extract"
    FILE_READ = "file-read"
    FILE_WRITE = "file-write"
    HTTP_REQUEST = "http-request"
    EMAIL_SEND = "email-send"
    SCRIPT = "script"


class NodePort(BaseModel):
    id: str
    name: str
    type: str
    required: bool
    description: Optional[str] = None


class WorkflowNode(BaseModel):
    id: str
    type: WorkflowNodeType
    name: str
    position: Dict[str, float]
    config: Dict[str, Any]
    inputs: List[NodePort]
    outputs: List[NodePort]


class EdgeCondition(BaseModel):
    type: str
    value: Any
    variable: Optional[str] = None


class WorkflowEdge(BaseModel):
    id: str
    source: str
    source_port: str
    target: str
    target_port: str
    label: Optional[str] = None
    condition: Optional[EdgeCondition] = None


class WorkflowVariable(BaseModel):
    name: str
    type: str
    value: Any
    description: Optional[str] = None
    is_input: bool
    is_output: bool


class WorkflowTrigger(BaseModel):
    id: str
    type: str
    config: Dict[str, Any]
    enabled: bool


class Workflow(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    status: WorkflowStatus
    nodes: List[WorkflowNode]
    edges: List[WorkflowEdge]
    variables: List[WorkflowVariable]
    triggers: List[WorkflowTrigger]
    created_at: datetime
    updated_at: datetime
    last_run_at: Optional[datetime] = None
    run_count: int = 0


class NodeExecution(BaseModel):
    node_id: str
    status: str
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    inputs: Dict[str, Any]
    outputs: Dict[str, Any]
    error: Optional[str] = None


class WorkflowRun(BaseModel):
    id: str
    workflow_id: str
    status: WorkflowRunStatus
    started_at: datetime
    completed_at: Optional[datetime] = None
    inputs: Dict[str, Any]
    outputs: Dict[str, Any]
    node_executions: List[NodeExecution]
    error: Optional[str] = None


class WorkflowCreateOptions(BaseModel):
    name: str
    description: Optional[str] = None
    nodes: List[WorkflowNode]
    edges: List[WorkflowEdge]
    variables: Optional[List[WorkflowVariable]] = None
    triggers: Optional[List[WorkflowTrigger]] = None


class WorkflowUpdateOptions(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    status: Optional[WorkflowStatus] = None
    nodes: Optional[List[WorkflowNode]] = None
    edges: Optional[List[WorkflowEdge]] = None
    variables: Optional[List[WorkflowVariable]] = None
    triggers: Optional[List[WorkflowTrigger]] = None


class WorkflowRunOptions(BaseModel):
    workflow_id: str
    inputs: Optional[Dict[str, Any]] = None
    async_exec: bool = True


class WorkflowListOptions(BaseModel):
    status: Optional[List[WorkflowStatus]] = None
    page: int = 1
    page_size: int = 20
    sort_by: str = "created_at"
    sort_order: str = "desc"
    search: Optional[str] = None


class WorkflowListResult(BaseModel):
    workflows: List[Workflow]
    total: int
    page: int
    page_size: int
    total_pages: int
